### PR TITLE
Add FXIOS-10609 [Native Error Page] Added flag to show no internet connection error page only

### PR DESCRIPTION
**Pull request by tusharC95**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10609)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/23236)

## :bulb: Description
- Added `no_internet_connection_error` in `nativeErrorPageFeature.yaml`. This parameter is exclusively to show/hide new "no internet connection" error page. 
- Additionally added a check for no internet connection error code.
- Added default strings for error page.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift [hunk 1]
```diff
@@ -789,13 +789,32 @@ extension BrowserViewController: WKNavigationDelegate {
...
+            guard var errorURLpath = URLComponents(string: "\(InternalURL.baseUrl)/\(ErrorPageHandler.path)" ) else { return }
+            errorURLpath.queryItems = [URLQueryItem(
+                name: InternalURL.Param.url.rawValue,
+                value: url.absoluteString
+            ), URLQueryItem(
+                name: "code",
+                value: String(
+                    error.code
+                )
+            )]
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:801]:
```suggestion
           errorURLpath.queryItems = [
                URLQueryItem(
                    name: InternalURL.Param.url.rawValue,
                    value: url.absoluteString
                ),
                URLQueryItem(
                    name: "code",
                    value: String(
                        error.code
                    )
                )
            ]
```

For arrays, suggested spacing is above. It's easier to read if each item is given it's own new line, with the wrapping `[` and `]` on their own lines.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift [hunk 2]
```diff
@@ -789,13 +789,32 @@ extension BrowserViewController: WKNavigationDelegate {
         }
 
         if let url = error.userInfo[NSURLErrorFailingURLErrorKey] as? URL {
+            guard var errorURLpath = URLComponents(string: "\(InternalURL.baseUrl)/\(ErrorPageHandler.path)" ) else { return }
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:792]: There are two places in this code where we `return` from a guard... but I wonder if we want to do that?

It looks like if `isNativeErrorPageEnabled` and `isNICErrorPageEnabled` are both false, we fall back to this line:
```
ErrorPageHelper(certStore: profile.certStore).loadPage(error, forUrl: url, inWebView: webView)
```

So if we can't get a `errorURLpath` or `errorPageURL`, should we also fallback to the `ErrorPageHelper`? 🤔 

(I might just be missing some context for your error page work)

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 3]
```diff
@@ -1537,10 +1542,16 @@ class BrowserViewController: UIViewController,
             return
         }
 
+        /// Used for checking if current error code is for no internet connection
+        let validNICError = url?.absoluteString.contains(String(Int(
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:1546]:
```suggestion
        let isValidNICError = url?.absoluteString.contains(String(Int(
```
Boolean naming

File: firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift [hunk 4]
```diff
@@ -153,11 +153,33 @@ class ErrorPageHandler: InternalSchemeResponse, FeatureFlaggable {
...
+        /// Used for checking if current error code is for no internet connection
+        let noInternetErrorCode = Int(
+            CFNetworkErrors.cfurlErrorNotConnectedToInternet.rawValue
+        )
+
+        if isNativeErrorPageEnabled {
+            return responseForNativeErrorPage(request: request)
+        } else if isNICErrorPageEnabled && (errCode == noInternetErrorCode) {
+            return responseForNativeErrorPage(request: request)
         } else {
```
Comment (ih-codes) [inline firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Hmmm, I'm concerned about this check. If we enabled native error pages, we will NEVER show your no internet connection error page, right? 🤔 

I think we need a little bit more logic here to determine the type of error pages we need to show, and THEN we can decide if we are allowed to show them (isNativeErrorPageEnabled, isNICErrorPageEnabled)

**Review by ih-codes — COMMENTED** [review]
Hi @tusharC95! 😄  I took a look and made a few swift best practices nitpicks.

I also raised a couple questions about some edge cases we might want to consider. 🤔 

If it sounds like I'm missing any context on this project, feel free to DM me to fill me in! Happy to discuss or answer questions too.

Great work, the fox is so cute on the no internet screen! 🥳

File: firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift [hunk 5]
```diff
@@ -74,6 +74,13 @@ final class FeatureFlagsDebugViewController: SettingsTableViewController, Featur
                 ) { [weak self] _ in
                     self?.reloadView()
                 },
+                FeatureFlagsBoolSetting(
+                    with: .noInternetConnectionErrorPage,
+                    titleText: format(string: "Enable NIC Native Error Page"),
+                    statusText: format(string: "Toggle to display natively created no internet conneciton error page")
```
[the remark this answers is not in the document]
Comment (tusharC95) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift:80]: Good Catch! I will fix this ASAP.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (tusharC95) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:792]: Agreed. I will rename it to errorPageURLComponents.

File: firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift (hunk 4 above)
Comment (ih-codes) [reply firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Oh, hold on, I just realized both these paths call the same function!

So maybe the if statements should just be combined in that case to be clear they are the same code paths? 🤔

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift (hunk 2 above)
Comment (tusharC95) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:792]: if isNativeErrorPageEnabled = false and isNICErrorPageEnabled = false, we should show the old error page, hence we fall back to ErrorPageHelper.

Comment (ih-codes) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:792]: Ok that makes sense!

So what happens if either `URLComponents(string: "\(InternalURL.baseUrl)/\(ErrorPageHandler.path)"` or `errorURLpath.url` return nil? 

It looks like the `guard` `else` just does a `return` in those 2 cases. Should we somehow return `ErrorPageHelper` instead? (Might need a little restructuring)

I do realize that these guard returns should never be called. 😅  So it's not a very important edge case. If you think it makes sense to leave the code like this, I'm fine with that. 👍

File: firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift (hunk 4 above)
Comment (tusharC95) [reply firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Good question!
If isNativeErrorPageEnabled = true then all the new error pages can be shown based on the error code (including "no internet" page). But if isNICErrorPageEnabled = true and error code is noInternetErrorCode, then only new "no internet error" page is shown. The current plan is to get No Internet Error Pages in v134, hence this temporary flag.

Comment (ih-codes) [reply firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Ahh I understand, thank you @tusharC95! This totally makes sense. 😄  You can disregard my comment here. 👍 

So the no internet error pages are going to be released first in v134, and the new error pages will come in a later release?

Comment (tusharC95) [reply firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Yes, that's the plan. 🤞

Comment (ih-codes) [reply firefox-ios/Client/Frontend/InternalSchemeHandler/ErrorPageHelper.swift:181]: Awesome, thank you for filling me in!

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift (hunk 2 above)
Comment (tusharC95) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Extensions/BrowserViewController+WebViewDelegates.swift:792]: Oh! Got your point. I will restructure the logic.

**Comment by tusharC95** [comment]
Thank you for the thorough testing 🫡. This issue could be linked to https://github.com/mozilla-mobile/firefox-ios/pull/23269. specifically BrowserViewType.nativeErrorPage. I will try to repro the bug and fix it asap.

**Review by ih-codes — APPROVED** [review]
Looks like your fix did the trick! 🎉

**Pull request merged**